File: bench/alpha/alpha.go
// alpha.go — alpha compositing (constant-opacity blend). Bit-identical to alpha.js.
package main

import (
	"fmt"
	"time"
)

const (
	W        = 512
	H        = 512
	N        = W * H * 4
	A        = 160
	IA       = 255 - A
	N_RUNS   = 21
	N_WARMUP = 5
)

func mix(h, x uint32) uint32 { return (h ^ x) * 0x01000193 }

func checksumU8(out []uint8) uint32 {
	h := uint32(0x811c9dc5)
	for _, b := range out {
		h = mix(h, uint32(b))
	}
	return h
}

func medianUs(samples []float64) uint64 {
	for i := 1; i < len(samples); i++ {
		v := samples[i]
		j := i - 1
		for j >= 0 && samples[j] > v {
			samples[j+1] = samples[j]
			j--
		}
		samples[j+1] = v
	}
	return uint64(samples[(len(samples)-1)>>1] * 1000.0)
}

func mkImage(out []uint8, seed uint32) {
	s := seed
	for i := range out {
		s ^= s << 13
		s ^= s >> 17
		s ^= s << 5
		out[i] = uint8(s & 255)
	}
}

func blend(src, dst, out []uint8) {
	for i := 0; i < N; i++ {
		out[i] = uint8((int(src[i])*A + int(dst[i])*IA + 127) >> 8)
	}
}

func main() {
	src := make([]uint8, N)
	dst := make([]uint8, N)
	out := make([]uint8, N)
	mkImage(src, 0x1234abcd)
	mkImage(dst, 0x7e1f93b5)

	for i := 0; i < N_WARMUP; i++ {
		blend(src, dst, out)
	}
	samples := make([]float64, N_RUNS)
	for i := 0; i < N_RUNS; i++ {
		t0 := time.Now()
		blend(src, dst, out)
		samples[i] = float64(time.Since(t0).Nanoseconds()) / 1e6
	}
	fmt.Printf("median_us=%d checksum=%d samples=%d stages=%d runs=%d\n",
		medianUs(samples), checksumU8(out), N, 1, N_RUNS)
}

File: bench/alpha/alpha.rs
// alpha.rs — alpha compositing (constant-opacity blend). Bit-identical to alpha.js.
use std::time::Instant;

const W: usize = 512;
const H: usize = 512;
const N: usize = W * H * 4;
const A: i32 = 160;
const IA: i32 = 255 - A;
const N_RUNS: usize = 21;
const N_WARMUP: usize = 5;

fn mix(h: u32, x: u32) -> u32 {
    (h ^ x).wrapping_mul(0x0100_0193)
}

fn checksum_u8(out: &[u8]) -> u32 {
    let mut h = 0x811c_9dc5u32;
    for &b in out {
        h = mix(h, b as u32);
    }
    h
}

fn median_us(samples: &mut [f64]) -> u64 {
    for i in 1..samples.len() {
        let v = samples[i];
        let mut j = i;
        while j > 0 && samples[j - 1] > v {
            samples[j] = samples[j - 1];
            j -= 1;
        }
        samples[j] = v;
    }
    (samples[(samples.len() - 1) >> 1] * 1000.0) as u64
}

fn mk_image(out: &mut [u8], seed: u32) {
    let mut s = seed;
    for x in out.iter_mut() {
        s ^= s << 13;
        s ^= s >> 17;
        s ^= s << 5;
        *x = (s & 255) as u8;
    }
}

fn blend(src: &[u8], dst: &[u8], out: &mut [u8]) {
    for i in 0..N {
        out[i] = (((src[i] as i32) * A + (dst[i] as i32) * IA + 127) >> 8) as u8;
    }
}

fn main() {
    let mut src = vec![0u8; N];
    let mut dst = vec![0u8; N];
    let mut out = vec![0u8; N];
    mk_image(&mut src, 0x1234abcd);
    mk_image(&mut dst, 0x7e1f93b5);

    for _ in 0..N_WARMUP {
        blend(&src, &dst, &mut out);
    }
    let mut samples = [0.0f64; N_RUNS];
    for s in &mut samples {
        let t0 = Instant::now();
        blend(&src, &dst, &mut out);
        *s = t0.elapsed().as_secs_f64() * 1000.0;
    }
    println!(
        "median_us={} checksum={} samples={} stages={} runs={}",
        median_us(&mut samples),
        checksum_u8(&out),
        N,
        1,
        N_RUNS
    );
}
